### PR TITLE
Reorganize OSX App

### PR DESCRIPTION
**Pull request by patternoia**
In this PR several things are addressed:
* The frameworks and dynamic libraries are moved from `Resources` to `Frameworks` to correspond to the canonical app bundle structure (as per [documentation](https://developer.apple.com/library/content/documentation/CoreFoundation/Conceptual/CFBundles/BundleTypes/BundleTypes.html#//apple_ref/doc/uid/10000123i-CH101-SW13))
* The app binary is moved to MacOS directory where it should reside
* The app binary is updated by `make` to look for libraries and frameworks in `@rpath` list enabling the relative paths (preferred way), rather than relying on launching script setting `DYLD_LIBRARY_PATH` and `DYLD_FRAMEWORK_PATH` env variables (not recommended).
* The game upon launch changes its working directory to the executable directory using [argv[0]](http://en.cppreference.com/w/cpp/language/main_function) (the `--basepath` option is still respected though), instead not launching. On macOS it checks the `../Resources/` dir existence and changes there, since `data` dir resides there.
* The launching script is updated accordingly.
* Linux launching script might profit from these changes as well.
* Minor: removed `Cocoa.framework` from linking list, since there is no dependence on it.

This PR provides the basis for an Xcode project for the macOS (ready, will follow in separate PR) and in future for official (long requested) iOS port for iPhone and iPad, since the author of these does not update his version and does not respond to any contact requests.

@remyroy and @BrettDong, you could help test this as macOS users.

### CODE REVIEW

**Comment by BrettDong** [comment]
Sorry, recently I switched to a Windows device.

**Comment by Leland** [comment]
Builds and runs just fine on my Mac. Am running 10.12.1 and build with Homebrew.

**Comment by patternoia** [comment]
Thanks for testing. Which scenario did you try? With SDL libs from Homebrew or with SDL frameworks?

**Comment by Leland** [comment]
Libs from Homebrew. I can get it built with installed SDL frameworks as well if need be. The `make` command I build with is as follows:


```bash
make app NATIVE=osx OSX_MIN=10.8 RELEASE=0 TILES=1 LOCALIZE=0 LUA=1 CLANG=1 CCACHE=1 USE_HOME_DIR=1 SOUND=1
```

Games runs as expected, and I can see the frameworks and libraries in `Cataclysm.app/Contents/Frameworks`

**Comment by patternoia** [comment]
Thanks for confirmation and brief response.

File: src/filesystem.cpp [hunk 1]
```diff
@@ -18,6 +18,7 @@
 #ifdef _MSC_VER
 #   include "wdirent.h"
 #   include <direct.h>
+#   define chdir _chdir
```
Comment (kevingranade) [inline src/filesystem.cpp]: I must be missing something, why is this being adjusted for VS builds in an OSX PR?

**Comment by kevingranade** [comment]
Sorry it took me a while to look at this, I was waiting on a confirmation of the SDL build, but I guess that's not happening.

Based on the stated goal of the PR, I expect to only see changes that will impact OSX builds, but I see changes to both Visual Studio and linux builds in here.  If OSX needs custom code, it needs to be behind explicit build flags.

File: src/filesystem.cpp (hunk 1 above)
Comment (patternoia) [inline src/filesystem.cpp]: This definition for MSVC is there to workaround the fact that posix (`chdir`) and windows (`_chdir`) directory change functions are not named the same. So this `#define` creates an alias under windows, which is used later in the code.
I've been referring to this [answer on SO](http://stackoverflow.com/a/40230786)  when introducing this change

Comment (kevingranade) [inline src/filesystem.cpp]: The problem is, we shouldn't be invoking this code except on OSX, we don't need this change on linux or windows.

**Comment by Leland** [comment]
@patternoia no rush of course, but I would love to see this updated & merged at some point!

**Comment by kevingranade** [comment]
No response to requests for correction in nearly a month, specifically changes to support OSX builds need to be isolated to OSX-specific code, in other words it should not change generic code.
Feel free to reopen if this issue is addressed.

**Pull request closed without merging**